packages/core/src/resolver.ts

```diff
@@ -23,37 +23,37 @@ type PipeFn<Prev, Next, PrevCtx extends Ctx, NextCtx = PrevCtx> = (
   c: PrevCtx,
 ) => Next extends ResultWithContext ? never : Next | ResultWithContext<Next, NextCtx>
 
-function pipe<A, Z>(ab: (i: A, c: Ctx) => Z): (input: A, ctx: Ctx) => Z
+function pipe<A, Z>(ab: (i: A, c: Ctx) => Z): (input: A, ctx: Ctx) => EnsurePromise<Z>
 function pipe<A, B, C, CA = Ctx, CB = CA, CC = CB>(
   ab: PipeFn<A, B, CA, CB>,
   bc: PipeFn<B, C, CB, CC>,
-): (input: A, ctx: CC) => C
+): (input: A, ctx: CA) => EnsurePromise<C>
```

Comment:
The core issue here was the type of `ctx` in this return function. Before we used the ctx type of the **last** function in the pipe. But this fixes it to instead be the ctx type of the **first** function in the pipe.

And then `EnsurePromise` fixes the return type in cases where the resolver pipe doesn't have an async function (resolver.pipe always returns a promise regardless of inputs)

Reply:
Does `EnsurePromise` truly match the runtime here? Doesn't that async function inside of `pipe` actually create a new `Promise` while EnsurePromise would preserve your "custom promises" (e.g. Bluebird).

```
interface MyPromise<T> extends PromiseLike<T> {
    abort(): void;
}

// MyPromise<number>
type result = EnsurePromise<MyPromise<number>>
```

Reply:
Currently we don't support custom promises. 

Without this conditional when you have an async function in the pipe, then the return type is `Promise<Promise<Result>>` which breaks types in the frontend with useQuery